### PR TITLE
pkg/apis: Add metric relabeling to ScrapeConfigs

### LINKED ISSUE #5803: Add metric_relabel_configs to ScrapeConfig CRD
**What is missing?**

Support of `metric_relabel_configs` directive in ScrapeConfig CRD.

Prometheus configuration : https://prometheus.io/docs/prometheus/latest/configuration/configuration/#metric_relabel_configs

**Why do we need it?**

To allow relabeling on samples before ingestion.

Today the ScrapeConfig CRD only supports relabeling before scraping (`relabelings` property)

**Environment**

* Prometheus Operator version: v0.67.1

**Anything else we need to know?**:

### PR DESCRIPTION
**Pull request by rimitchell**
## Description

Brings support for metric relabeling to the ScrapeConfig object. Aligning with the existing functionality of Service, Pod and Probe monitors.

Fixes #5803 

## Type of change

_What type of changes does your code introduce to the Prometheus operator? Put an `x` in the box that apply._

- [ ] `CHANGE` (fix or feature that would cause existing functionality to not work as expected)
- [x] `FEATURE` (non-breaking change which adds functionality)
- [ ] `BUGFIX` (non-breaking change which fixes an issue)
- [ ] `ENHANCEMENT` (non-breaking change which improves existing functionality)
- [ ] `NONE` (if none of the other choices apply. Example, tooling, build system, CI, docs, etc.)

## Changelog entry

ScrapeConfigs: Now supports metric relabeling

```release-note
- Add metricRelabel functionality to ScrapeConfigs
```

### CODE REVIEW

File: pkg/prometheus/promcfg.go
Comment (xiu) [inline pkg/prometheus/promcfg.go:2524]: Most ScrapeConfig unit tests are failing as we're passing a new empty `metric_relabel_config` field to the configuration if it's not set. The test cases you updated added the key but not at the right spot. The following ensures that we're adding the key only if necessary.

```suggestion
      if sc.Spec.MetricRelabelConfigs != nil {
	cfg = append(cfg, yaml.MapItem{Key: "metric_relabel_configs", Value: generateRelabelConfig(labeler.GetRelabelingConfigs(sc.TypeMeta, sc.ObjectMeta, sc.Spec.MetricRelabelConfigs))})
      }
```

File: pkg/prometheus/promcfg_test.go
Comment (xiu) [inline pkg/prometheus/promcfg_test.go:8882]: We shouldn't add an empty slice if `spec.MetricRelabelConfigs` wasn't set, see my previous comment. This is true for all the tests were you added this.

**Review by xiu — CHANGES_REQUESTED** [review]
Thanks for this! Added some comments. To regenerate assets, please use `make generate` and commit the results.

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go
Comment (nicolastakashi) [inline pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:148]: Small nit: we should include +optional annotation

File: pkg/prometheus/promcfg.go
Comment (rimitchell) [reply pkg/prometheus/promcfg.go]: Oops. My bad on missing the generated files. I've added this check and the optional annotation below and have now included the generated files

File: pkg/prometheus/promcfg.go
Comment (nicolastakashi) [inline pkg/prometheus/promcfg.go:2523]: I'd suggest this just to improve a bit readability.
```suggestion
	if sc.Spec.MetricRelabelConfigs != nil {
		relabelConfigValue = generateRelabelConfig(labeler.GetRelabelingConfigs(sc.TypeMeta, sc.ObjectMeta, sc.Spec.MetricRelabelConfigs))
		cfg = append(cfg, yaml.MapItem{
			Key:   "metric_relabel_configs",
			Value: relabelConfigValue,
		})
	}	
```

File: pkg/prometheus/resource_selector.go
Comment (ArthurSens) [reply pkg/prometheus/resource_selector.go:694]: Option 1 isn't good since we'd skip the rejection function if the `action` field is empty.

I haven't looked deeply into the code, but option 3 could indeed be a change in functionality for other ServiceMonitors, PodMonitors and Probes. Let's keep that in a separate PR, if relevant 😅 

Option 2 looks like a good suggestion, but I can see that we're already using this same pattern every other place we call `validateRelabelConfig`. It would be nice to change those in a separate PR if you feel like it :)

**Review by ArthurSens — APPROVED** [review]
LGTM! Thanks for the PR

I'll let others take another look before merging it :)

File: pkg/prometheus/resource_selector.go
Comment (nicolastakashi) [reply pkg/prometheus/resource_selector.go:694]: Yeah!
It's totally fine for me keep as it is and we do the changes we think it's valuable in another PR 💪🏽

**Review by simonpasquier — COMMENTED** [review]
I suggest to wait for https://github.com/prometheus-operator/prometheus-operator/pull/5841 which is fixing a flaw in the current code regarding resource selection.

File: pkg/prometheus/resource_selector.go
Comment (ArthurSens) [inline pkg/prometheus/resource_selector.go:701]:
```suggestion
		if err = validateRelabelConfigs(rs.p, sc.Spec.MetricRelabelConfigs); err != nil {
			rejectFn(sc, fmt.Errorf("metricRelabelConfigs: %w", err))
			continue
		}
```

**Review by ArthurSens — COMMENTED** [review]
@rimitchell, could you rebase and make one adjustment?

Simon found and fixed a bug in #5841 that affects your PR 😬

File: pkg/prometheus/resource_selector.go
Comment (rimitchell) [reply pkg/prometheus/resource_selector.go]: Rebased and code updated to use the _validateRelabelConfigs_ function

File: pkg/prometheus/resource_selector.go
Comment (rimitchell) [reply pkg/prometheus/resource_selector.go:1]: Thank you for catching that. I've added a set for the metric relabeling

**Review by ArthurSens — APPROVED** [review]
Thanks for your patience through the whole process!

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go
Comment (nicolastakashi) [reply pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:149]: Very small nit but I'm not sure if we need a pointer here since default value for array in go lang is empty.
wdyt @ArthurSens and @simonpasquier ?

File: pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go
Comment (nicolastakashi) [inline pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:149]: I'm not sure if we need a pointer here since default value for an array in golang is empty.
wdyt @ArthurSens and @simonpasquier ?

**Review by simonpasquier — APPROVED** [review]
Thanks!

**Pull request merged**
